pkg/reconciler/cluster/startup.go

```diff
@@ -0,0 +1,128 @@
+	return &o
+}
+
+// Options are the options for the cluster controller
+type Options struct {
+	SyncerImage     string
+	PullMode        bool
+	PushMode        bool
+	AutoPublishAPIs bool
+	NumTreads       int
+	ResourcesToSync []string
+}
+
+func (o *Options) Validate() error {
+	if o.PullMode && o.PushMode {
+		return errors.New("can't set both --push_mode and --pull_mode")
+	}
+	return nil
+}
+
+func (o *Options) Start(ctx context.Context, kubeconfig clientcmdapi.Config, kcpSharedInformerFactory kcpexternalversions.SharedInformerFactory, crdSharedInformerFactory crdexternalversions.SharedInformerFactory) error {
+	// Register CRDs in both the admin and user logical clusters
+	requiredCrds := []metav1.GroupKind{
+		{Group: apiresourceapi.GroupName, Kind: "apiresourceimports"},
+		{Group: apiresourceapi.GroupName, Kind: "negotiatedapiresources"},
+		{Group: clusterapi.GroupName, Kind: "clusters"},
+	}
+	for _, contextName := range []string{"admin", "user"} {
+		logicalClusterConfig, err := clientcmd.NewNonInteractiveClientConfig(kubeconfig, contextName, &clientcmd.ConfigOverrides{}, nil).ClientConfig()
+		if err != nil {
+			return err
+		}
+		crdClient := apiextensionsv1client.NewForConfigOrDie(logicalClusterConfig).CustomResourceDefinitions()
+		if err := config.BootstrapCustomResourceDefinitions(ctx, crdClient, requiredCrds); err != nil {
+			return err
+		}
+	}
+
+	if o.PullMode && o.PushMode {
+		return errors.New("can't set --push_mode and --pull_mode")
+	}
+	syncerMode := SyncerModeNone
+	if o.PullMode {
+		syncerMode = SyncerModePull
+	}
+	if o.PushMode {
+		syncerMode = SyncerModePush
+	}
+
+	adminConfig, err := clientcmd.NewNonInteractiveClientConfig(kubeconfig, "admin", &clientcmd.ConfigOverrides{}, nil).ClientConfig()
+	if err != nil {
+		return err
+	}
+	clientutils.EnableMultiCluster(adminConfig, nil, true, "clusters", "customresourcedefinitions", "apiresourceimports", "negotiatedapiresources")
+
+	apiExtensionsClient := apiextensionsclient.NewForConfigOrDie(adminConfig)
+	kcpClient := kcpclient.NewForConfigOrDie(adminConfig)
+
+	clusterController, err := NewController(
+		apiExtensionsClient,
+		kcpClient,
+		kcpSharedInformerFactory.Cluster().V1alpha1().Clusters(),
+		kcpSharedInformerFactory.Apiresource().V1alpha1().APIResourceImports(),
+		o.SyncerImage,
+		kubeconfig,
+		o.ResourcesToSync,
+		syncerMode,
+	)
+	if err != nil {
+		return err
+	}
+
+	apiresourceController, err := apiresource.NewController(
+		apiExtensionsClient,
+		kcpClient,
+		o.AutoPublishAPIs,
+		kcpSharedInformerFactory.Apiresource().V1alpha1().NegotiatedAPIResources(),
+		kcpSharedInformerFactory.Apiresource().V1alpha1().APIResourceImports(),
+		crdSharedInformerFactory.Apiextensions().V1().CustomResourceDefinitions(),
+	)
+	if err != nil {
+		return err
+	}
+
+	kcpSharedInformerFactory.Start(ctx.Done())
+	crdSharedInformerFactory.Start(ctx.Done())
+	kcpSharedInformerFactory.WaitForCacheSync(ctx.Done())
```

Comment:
And if we don't start the factories here, we can't wait for cache sync directly here either. We can either put both the waiting and the controller start in a goroutine, or we can pass individual informer.Synced pointers into each controller, and wait on those inside each controller's Start func.